src/pages/instances/actions/snapshots/useInstanceSnapshot.tsx

```diff
@@ -0,0 +1,253 @@
+import {
+  Icon,
+  Input,
+  List,
+  Tooltip,
+  useNotify,
+} from "@canonical/react-components";
+import { useQueryClient } from "@tanstack/react-query";
+import {
+  createSnapshot,
+  renameSnapshot,
+  updateSnapshot,
+} from "api/instance-snapshots";
+import ItemName from "components/ItemName";
+import { useEventQueue } from "context/eventQueue";
+import { useFormik } from "formik";
+import React, { ReactNode, useState } from "react";
+import { LxdInstance, LxdInstanceSnapshot } from "types/instance";
+import {
+  UNDEFINED_DATE,
+  getBrowserFormatDate,
+  stringToIsoTime,
+} from "util/helpers";
+import { queryKeys } from "util/queryKeys";
+import {
+  SnapshotFormValues,
+  getExpiresAt,
+  getInstanceSnapshotSchema,
+} from "util/snapshots";
+import { TOOLTIP_OVER_MODAL_ZINDEX } from "util/zIndex";
+
+interface CreateInstanceSnapshotHook {
+  instance: LxdInstance;
+  close: () => void;
+  onSuccess: (message: ReactNode) => void;
+  isStateful: boolean;
+  isRunning: boolean;
+}
+
+interface EditInstanceSnapshotHook {
+  instance: LxdInstance;
+  snapshot: LxdInstanceSnapshot;
+  close: () => void;
+  onSuccess: (message: ReactNode) => void;
+}
+
+const useCreateInstanceSnapshot = (props: CreateInstanceSnapshotHook) => {
+  const { instance, close, onSuccess, isStateful, isRunning } = props;
+  const eventQueue = useEventQueue();
+  const notify = useNotify();
+  const queryClient = useQueryClient();
+  const controllerState = useState<AbortController | null>(null);
+
+  const formik = useFormik<SnapshotFormValues<{ stateful?: boolean }>>({
+    initialValues: {
+      name: "",
+      stateful: false,
+      expirationDate: null,
+      expirationTime: null,
+    },
+    validateOnMount: true,
+    validationSchema: getInstanceSnapshotSchema(instance, controllerState),
+    onSubmit: (values, { resetForm }) => {
+      notify.clear();
+      const expiresAt =
+        values.expirationDate && values.expirationTime
+          ? stringToIsoTime(
+              getExpiresAt(values.expirationDate, values.expirationTime),
+            )
+          : UNDEFINED_DATE;
+      void createSnapshot(
+        instance,
+        values.name,
+        expiresAt,
+        values.stateful || false,
+      ).then((operation) =>
+        eventQueue.set(
+          operation.metadata.id,
+          () => {
+            void queryClient.invalidateQueries({
+              predicate: (query) => query.queryKey[0] === queryKeys.instances,
+            });
+            onSuccess(
+              <>
+                Snapshot <ItemName item={values} bold /> created.
+              </>,
+            );
+            resetForm();
```

Comment:
The error handling does not surface the error message. I.e. when on a project that doesn't allow snapshot creation, the api responds with a message `Project \"default\" doesn't allow for snapshot creation`. We should show this message to the user. This is also relevant for the storage volume snapshot creation and edit snapshot on an instance or volume.

Reply:
I see, I didn't catch the edge case that if project is restricted after instance or volume snapshots were created, then the user can still create snapshots but this will fail. Will check why the error handling isn't showing the message here.